### PR TITLE
Add a way to specify default argument to action using variable dict a…

### PR DESCRIPTION
**Pull request by Yannig**
##### ISSUE TYPE
- Feature Pull Request
##### ANSIBLE VERSION

```
ansible 2.2.0 (devel_ansible_modules_args 56b40a5bb5) last updated 2016/06/04 15:18:04 (GMT +200)
  lib/ansible/modules/core: (detached HEAD ca4365b644) last updated 2016/06/02 19:40:09 (GMT +200)
  lib/ansible/modules/extras: (detached HEAD b0aec50b9a) last updated 2016/06/02 19:40:09 (GMT +200)
  config file = /etc/ansible/ansible.cfg
```
##### SUMMARY

This patch add a way to easily specify default values to any Ansible action in order to avoid to specify each time the same values.

To get an idea of what can be done with it, you can grab an example at the following location: https://github.com/Yannig/yannig-ansible-playbooks/tree/master/ansible_modules_args

Let's take the following inventory:

``` ini
host1
host2
host3

[all:vars]
ansible_connection=local
```

For the host **host1**, we will define the following default values:

``` yaml
ansible_module_args:
  copy:
    mode: "0624"
```

For **host2**, we will define the following default values:

``` yaml
ansible_module_args:
  copy:
    mode: "0644"
```

Now, let's play the following playbook:

``` yaml
- name: "Simple example"
  hosts: all
  gather_facts: no
  tasks:
    - copy: content={{inventory_hostname}} dest=/tmp/force-mode-{{  inventory_hostname}} mode=0600
    - copy: content={{inventory_hostname}} dest=/tmp/default-mode-{{inventory_hostname}}
    - shell: ls -l /tmp/*-mode-{{inventory_hostname}}
      register: _
      changed_when: no
    - debug: var=_.stdout_lines
```

You will get the following result:

```
TASK [debug] *******************************************************************
ok: [host1] => {
    "_.stdout_lines": [
        "-rw--w-r-- 1 yannig yannig 5 juin   4 15:15 /tmp/default-mode-host1", 
        "-rw------- 1 yannig yannig 5 juin   4 15:15 /tmp/force-mode-host1"
    ]
}
ok: [host2] => {
    "_.stdout_lines": [
        "-rw-r--r-- 1 yannig yannig 5 juin   4 15:15 /tmp/default-mode-host2", 
        "-rw------- 1 yannig yannig 5 juin   4 15:15 /tmp/force-mode-host2"
    ]
}
ok: [host3] => {
    "_.stdout_lines": [
        "-rw-rw-r-- 1 yannig yannig 5 juin   4 15:15 /tmp/default-mode-host3", 
        "-rw------- 1 yannig yannig 5 juin   4 15:15 /tmp/force-mode-host3"
    ]
}
```

You can see that forcing the mode option in copy module result in the same value for all. If you do not specify the value, you get a value driven by the ansible_module_args content.

Another good example could be to use this to specify http_proxy for every yum action. Here's an example:

``` yaml
ansible_module_args:
  yum:
    environment: { http_proxy: "http://127.0.0.1:8080" }
```

### CODE REVIEW

**Comment by sivel** [comment]
I am a -1 on this.  This feels too magical to me and has the ability to make things very complicated to understand and to know where the task is built and hwere data is coming from.

The current recommendation for task defaults is to use YAML anchors and aliases.  For one, the anchors/aliases way, requires you to still reference a "variable" in the task, whereas your implementation doesn't and basically allows the data to come from anywhere making things hard to follow.

**Comment by Yannig** [comment]
@sivel thanks for your feedback.

The idea behind this features is to reproduce the puppet resource default statement. You can have an example of this feature by following this url: https://docs.puppet.com/puppet/latest/reference/lang_defaults.html

What I found anoying is sometime have to specify sames options for every actions and have no way to specify a default behaviour for a given context.

**Comment by bcoca** [comment]
I would not do this through ansible.cfg as the play now will behave very differently depending on configuration files. 

I do think the idea might have some merit if made part of roles (meta/main.yml set default for current role?) <=  spitballing here

**Comment by sivel** [comment]
I would preferably like to see this be more explicit.

Such as allowing each task to accept a var that defines a set of defaults, instead of implicit magic, such as:

```
- copy: content={{inventory_hostname}} dest=/tmp/force-mode-{{  inventory_hostname}} mode=0600
  defaults: "{{ ansible_copy_args }}"
```

Or, since we already have the ability to use `args` for a task, maybe we could "re-use" that.

**Comment by Yannig** [comment]
@sivel I will follow your suggestion as it could fulfill the function I was thinking about.

**Comment by jctanner** [comment]
Duplicates https://github.com/ansible/ansible/pull/22648 which we have been debating and shaping in meetings for a while. Closing this in favor of that.
